jvb/src/main/kotlin/org/jitsi/videobridge/Endpoint.kt

```diff
@@ -1239,4 +1359,379 @@ class Endpoint @JvmOverloads constructor(
+                            timestampDelta =
+                                RtpUtils.getTimestampDiff(state.lastTimestamp, prevTimestamp)
+                            if (state.lastTl0Index != -1 && packet is Vp9Packet)
+                                tl0IndexDelta = (256 + state.lastTl0Index - (packet.TL0PICIDX - 1)) % 256
+                            else
+                                tl0IndexDelta = 0
+                        }
+                    }
+                    recv.hasDeltas = true
+                }
+
+                recv.state.update(packet)
+
+                packet.ssrc = ssrc
+                packet.sequenceNumber = RtpUtils.applySequenceNumberDelta(packet.sequenceNumber, sequenceNumberDelta)
+                packet.timestamp = RtpUtils.applyTimestampDelta(packet.timestamp, timestampDelta)
+                if (packet is Vp9Packet) {
+                    packet.TL0PICIDX = (packet.TL0PICIDX + tl0IndexDelta) % 256
+                }
+
+                state.update(packet)
+            } else {
+                /* Don't touch send state if we're dropping the packet. */
+                recv.state.update(packet)
+            }
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        override fun toString(): String = "$ssrc{$state,\u2206=$sequenceNumberDelta/$timestampDelta/$tl0IndexDelta}"
+    }
+
+    /**
+     * Associates primary and secondary send SSRCs.
+     */
+    private class SendSource(val props: SourceDesc, val send1: SendSsrc, val send2: SendSsrc) {
+
+        /**
+         * If false, do not send on this SSRC until a packet with start=true arrives.
+         */
+        private var started = false
+
+        constructor(props: SourceDesc) : this(props, SendSsrc(null), SendSsrc(null))
+
+        /**
+         * Demux to proper SSRC.
+         */
+        private fun getSender(ssrc: Long) = if (ssrc == props.ssrc2) send2 else send1
+
+        /**
+         * Update RTP state and apply deltas.
+         * Returns true if packet should be sent.
+         */
+        fun rewriteRtp(packet: RtpPacket, start: Boolean, recv: ReceiveSsrc): Boolean {
+            if (start) {
+                started = true
+            }
+            getSender(packet.ssrc).rewriteRtp(packet, started, recv)
+            return started
+        }
+
+        /**
+         * {@inheritDoc}
+         */
+        override fun toString(): String = "$send1, $send2" + if (started) "" else " (not started)"
+    }
+
+    /**
+     * Limit the number of local SSRCs used for the given media type to the number specified.
+     * If there are more sources in the conference than the limit, then the least recently used
+     * SSRCs are remapped. RTP packets have their header fields rewritten so the stream appears
+     * to be a continuation of an already advertised SSRC.
+     */
+    private inner class SsrcCache(val mediaType: MediaType, val size: Int) {
```

Comment:
Rather than having SsrcCache take a mediaType argument and then have if(mediaType) switches in its code, it seems like it would be cleaner to have an abstract SsrcCache with AudioSsrcCache and VideoSsrcCache subclasses.

Reply:
Yes, seems much better now. thanks for the suggestion.